showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITGdch.java

```diff
@@ -0,0 +1,183 @@
+  private static final String CA_CERT_JSON_KEY = "ca_cert_path";
+  private static final String TEMP_CREDENTIAL_JSON_FILENAME = "temp_gdch_credential.json";
+
+  @Rule public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  private EchoClient client;
+  private EchoSettings settings;
+  private Credentials credentials;
+
+  @Before
+  public void setup() throws IOException, URISyntaxException {
+    prepareCredentials();
+    client = null;
+    tempFolder.create();
+    settings =
+        EchoSettings.newBuilder()
+            .setCredentialsProvider(FixedCredentialsProvider.create(credentials))
+            .build();
+  }
+
+  private void prepareCredentials() throws IOException, URISyntaxException {
+    // compute absolute path of the CA certificate
+    Path caCertPath = Paths.get(getClass().getResource(CA_CERT_RESOURCE_PATH).toURI());
+
+    // open gdch credential json (still needs its "ca_cert_path" to point to the CA certificate
+    // obtained from above)
+    JsonFactory factory = new GsonFactory();
+    GenericJson converted =
+        factory.fromInputStream(
+            getClass().getResourceAsStream(TEST_GDCH_CREDENTIAL_FILE), GenericJson.class);
+
+    // modify and save to a temporary folder
+    converted.set(CA_CERT_JSON_KEY, caCertPath.toAbsolutePath().toString());
+    File tempGdchCredentialFile = tempFolder.newFile(TEMP_CREDENTIAL_JSON_FILENAME);
+    try (FileWriter fileWriter = new FileWriter(tempGdchCredentialFile)) {
+      String preparedJson = converted.toPrettyString();
+      fileWriter.write(preparedJson);
+    }
+
+    // use temp location to instantiate credentials
+    credentials = GdchCredentials.fromStream(new FileInputStream(tempGdchCredentialFile));
+  }
+
+  /**
+   * {@link com.google.api.gax.rpc.ClientContext} sets a credentials object many times before
+   * concluding its creation. double check that they end up being the same
+   */
+  @Test
+  public void testClientWithGdchCredential_keepsCredentials() throws IOException {
+    Exception unexpected = getExceptionFromClientCreation();
+    assertNull(unexpected);
+    assertSame(credentials, client.getSettings().getCredentialsProvider().getCredentials());
+  }
+
+  /**
+   * Confirms setting a valid audience is successful.
+   *
+   * The provider's credentials do not get altered when adding audience.
+   * They are recreated and used only inside `ClientContext`
+   * @throws IOException
+   * @throws URISyntaxException
+   */
+  @Test
+  public void testClientWithGdchCredentialWithValidAudience_correct()
+      throws IOException, URISyntaxException {
+    String audience = "valid-audience";
+    settings = settings.toBuilder().setGdchApiAudience(audience).build();
+    Exception unexpected = getExceptionFromClientCreation();
+    assertNull(unexpected);
+    GdchCredentials fromClient =
+            (GdchCredentials) client.getSettings().getCredentialsProvider().getCredentials();
+
+    assertSame(credentials, fromClient);
```

Comment:
I thought the credential would be different since it now has an additional field `gdchApiAudience`?

Reply:
So the provider is not modified when creating the `ClientContext`. In fact, `ClientContext.create()` retrieves the `Credentials` from the provider, then modifies them accordingly (e.g. re-create with `gdchApiAudience` or any other type of `Credentials) and stores them as an internal property. 

I'm just describing what happens with `ClientContext` creation. 
For example, 
https://github.com/googleapis/sdk-platform-java/blob/ea05922a7010a274db9e38be3940cc54a02a585b/gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L193-L199

Modifies the `Credentials` object and then a new `ClientContext` is created with such `Credentials`

https://github.com/googleapis/sdk-platform-java/blob/ea05922a7010a274db9e38be3940cc54a02a585b/gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L261-L264

Should `ClientContext.create()` alter the `Credentials` provided by `CredentialsProvider`?

Reply:
While the logic above is correct, the test seems not validating what it declares (Blake's point from different angle). The test description WithAudience - is is not tested. 

Instead, test validates that credentials stay in credential provider as is. 

This is effectively a unit test, not integration

Reply:
> Also, this is effectively a unit test, not integration, right?

We want to test that the audience set through `ClientSetting` in a client library is being properly propagated to `ClientContext` and set to `Credentials`, which we cannot test through unit tests, only through showcase clients. So yeah this is not a true integration tests that interact with servers, more like a component test only verifying the client behaviors.
That being said, I agree with you that this test does not verify the audience, which we should.